metricplugin/wiretap.go

```diff
@@ -23,17 +27,177 @@ var (
+// receive duplicate disconnection events or miss connection events.
+// However, it looks like we converge with the true IPFS connectivity after a
+// while.
 type BitswapWireTap struct {
-	api         *core.IpfsNode
+	api       *core.IpfsNode
+	bsnetImpl bsnet.BitSwapNetwork
+
+	// Keeps track of whom to send Bitswap messages and connection events to.
 	subscribers subscribers
+
+	// Keeps track of peers we are connected to and bitswap senders.
+	connManager connectionManager
+
+	// Lifetime management.
+	closing   chan struct{}
+	closeLock sync.Mutex
+	wg        sync.WaitGroup
+}
+
+// Tracks which peers we are connected to, and whether we have a Bitswap sender
+// to them.
+type connectionManager struct {
+	lock           sync.Mutex
+	allConnections map[string]struct{}
+	peers          map[peer.ID]peerConnection
+}
+
+// Tracks connections to a single peer and whether we have a Bitswap sender to
+// this peer.
+type peerConnection struct {
+	bitswapSender bsnet.MessageSender
+	connections   map[string]network.Conn
 }
 
 type subscribers struct {
 	subscribers []EventSubscriber
 	lock        sync.RWMutex
 }
 
+// NewWiretap creates a new Bitswap Wiretap and network Notifiee.
+// This will spawn goroutines to log stats and connect to peers via Bitswap.
+// Use the Shutdown method to shut down cleanly.
+func NewWiretap(ipfsInstance *core.IpfsNode, bsnetImpl bsnet.BitSwapNetwork) *BitswapWireTap {
+	wt := &BitswapWireTap{
+		api:       ipfsInstance,
+		bsnetImpl: bsnetImpl,
+		connManager: connectionManager{
+			allConnections: make(map[string]struct{}),
+			peers:          make(map[peer.ID]peerConnection),
+		},
+	}
+
+	// Start a goroutine to periodically go through all peers and open Bitswap
+	// senders if we're missing any.
+	wt.wg.Add(1)
+	go wt.bitswapConnectionLoop()
+
+	// Start a goroutine to periodically print stats about our perceived
+	// connectivity and Bitswap senders.
+	// This also populates prometheus.
+	wt.wg.Add(1)
+	go wt.connectivityStatLoop()
+
+	return wt
+}
+
+// WaitGroup-bound function to loop forever and print stats about connectivity
+// and Bitswap senders.
+// This exits when wt.closing is closed.
+func (wt *BitswapWireTap) connectivityStatLoop() {
+	defer wt.wg.Done()
+
+	// A list of Bitswap protocol IDs.
+	// These are the official IPFS implementations of Bitswap.
+	// There are more (e.g., ipfs-embed), but we can't speak those.
+	bitswapProtocols := map[protocol.ID]struct{}{
+		bsnet.ProtocolBitswapNoVers:  {},
+		bsnet.ProtocolBitswapOneZero: {},
+		bsnet.ProtocolBitswapOneOne:  {},
+		bsnet.ProtocolBitswap:        {},
+	}
+
+	// Ticker for periodic printing/prometheus populating.
+	ticker := time.NewTicker(time.Duration(10) * time.Second)
+
+	for {
+		select {
+		case <-wt.closing:
+			log.Debug("Wiretap connectivity stat loop exiting.")
+			return
+		case <-ticker.C:
+		}
+
+		// Get stats as reported by the IPFS node.
+		conns := wt.api.PeerHost.Network().Conns()
```

Comment:
1) Do you know if this locks the `Network` and how long it takes?
2) Rust-inspired thought: Do we have to make sure not to accidentally alter the connections when storing them in the map?

Reply:
This does read-lock the connection management thing in the Network implementation, but only to copy out all the connections. From debug logging in the `populatePrometheus` method: this doesn't take long.

The connections we get are copies, but `network.Conn` is an interface, so we're actually probably getting copies of pointers. This means that yes, if we do stuff with the connections, we're probably working on the same instance as the IPFS node. We're not doing anything with them (except storing, which only affects garbage collection). The `bsnet.MessageSender` implementation asks the IPFS node for a connection+stream to the given peer ID, so that's also independent of what we're doing here.